File: packages/firebase_ml_vision/ios/Classes/LandmarkDetector.m
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FLTFirebaseMlVisionPlugin.h"

@interface LandmarkDetector ()
@property FIRVisionCloudLandmarkDetector *detector;
@end

@implementation LandmarkDetector
- (instancetype)initWithVision:(FIRVision *)vision options:(NSDictionary *)options {
  self = [super init];
  if (self) {
    _detector = [vision cloudLandmarkDetectorWithOptions:[LandmarkDetector parseOptions:options]];
  }
  return self;
}

- (void)handleDetection:(FIRVisionImage *)image result:(FlutterResult)result {
  [_detector
      detectInImage:image
        completion:^(NSArray<FIRVisionCloudLandmark *> *_Nullable landmarks, NSError *_Nullable error) {
          if (error) {
            [FLTFirebaseMlVisionPlugin handleError:error result:result];
            return;
          } else if (!landmarks) {
            result(@[]);
            return;
          }

          NSMutableArray *landmarkData = [NSMutableArray array];
          for (FIRVisionCloudLandmark *landmark in landmarks) {
            NSArray<FIRVisionLatitudeLongitude *> *visionLocations = landmark.locations;
            NSMutableArray *locations = [[NSMutableArray alloc] initWithCapacity:[visionLocations count]];
            for (int i = 0; i < [visionLocations count]; i++) {
              FIRVisionLatitudeLongitude *location = [visionLocations objectAtIndex:i];
              NSDictionary *locationDictionary = @{
                @"lat" : location.latitude,
                @"lng" : location.longitude
              };
              [locations insertObject:locationDictionary atIndex:i];
            }

            NSDictionary *data = @{
              @"left" : @(landmark.frame.origin.x),
              @"top" : @(landmark.frame.origin.y),
              @"width" : @(landmark.frame.size.width),
              @"height" : @(landmark.frame.size.height),
              @"confidence" : landmark.confidence,
              @"entityId" : landmark.entityId,
              @"landmark" : landmark.landmark,
              @"locations" : locations
            };

            [landmarkData addObject:data];
          }

          result(landmarkData);
        }];
}

+ (FIRVisionCloudDetectorOptions *)parseOptions:(NSDictionary *)optionsData {
  FIRVisionCloudDetectorOptions *options = [[FIRVisionCloudDetectorOptions alloc] init];
  NSNumber *maxResults = optionsData[@"maxResults"];
  options.maxResults = [maxResults integerValue];

  NSString *mode = optionsData[@"modelType"];
  options.modelType = FIRVisionCloudModelTypeStable;
  if ([mode isEqualToString:@"latest_model"]) {
    options.modelType = FIRVisionCloudModelTypeLatest;
  }

  return options;
}
@end

File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/LandmarkDetector.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebasemlvision;

import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.cloud.FirebaseVisionCloudDetectorOptions;
import com.google.firebase.ml.vision.cloud.landmark.FirebaseVisionCloudLandmark;
import com.google.firebase.ml.vision.cloud.landmark.FirebaseVisionCloudLandmarkDetector;
import com.google.firebase.ml.vision.common.FirebaseVisionImage;
import com.google.firebase.ml.vision.common.FirebaseVisionLatLng;
import io.flutter.plugin.common.MethodChannel;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class LandmarkDetector implements Detector {
  private final FirebaseVisionCloudLandmarkDetector detector;

  LandmarkDetector(FirebaseVision vision, Map<String, Object> options) {
    detector = vision.getVisionCloudLandmarkDetector(parseOptions(options));
  }

  @Override
  public void handleDetection(final FirebaseVisionImage image, final MethodChannel.Result result) {
    detector
        .detectInImage(image)
        .addOnSuccessListener(
            new OnSuccessListener<List<FirebaseVisionCloudLandmark>>() {
              @Override
              public void onSuccess(
                  List<FirebaseVisionCloudLandmark> firebaseVisionCloudLandmarks) {
                List<Map<String, Object>> landmarks =
                    new ArrayList<>(firebaseVisionCloudLandmarks.size());
                for (FirebaseVisionCloudLandmark landmark : firebaseVisionCloudLandmarks) {
                  Map<String, Object> landmarkData = new HashMap<>();

                  landmarkData.put("left", (double) landmark.getBoundingBox().left);
                  landmarkData.put("top", (double) landmark.getBoundingBox().top);
                  landmarkData.put("width", (double) landmark.getBoundingBox().width());
                  landmarkData.put("height", (double) landmark.getBoundingBox().height());

                  landmarkData.put("confidence", landmark.getConfidence());
                  landmarkData.put("entityId", landmark.getEntityId());
                  landmarkData.put("landmark", landmark.getLandmark());

                  List<Map> locations = new ArrayList<>();
                  for (FirebaseVisionLatLng latLng : landmark.getLocations()) {
                    HashMap<String, Double> location = new HashMap<>();
                    location.put("lat", latLng.getLatitude());
                    location.put("lng", latLng.getLongitude());
                    locations.add(location);
                  }
                  landmarkData.put("locations", locations);

                  landmarks.add(landmarkData);
                }

                result.success(landmarks);
              }
            })
        .addOnFailureListener(
            new OnFailureListener() {
              @Override
              public void onFailure(@NonNull Exception exception) {
                result.error("landmarkDetectorError", exception.getLocalizedMessage(), null);
              }
            });
  }

  private FirebaseVisionCloudDetectorOptions parseOptions(Map<String, Object> options) {
    Integer maxResults = (Integer) options.get("maxResults");

    int model = FirebaseVisionCloudDetectorOptions.STABLE_MODEL;
    if (((String) options.get("modelType")).equals("lastest_model")) {
      model = FirebaseVisionCloudDetectorOptions.LATEST_MODEL;
    }

    FirebaseVisionCloudDetectorOptions.Builder builder =
        new FirebaseVisionCloudDetectorOptions.Builder()
            .setMaxResults(maxResults)
            .setModelType(model);

    return builder.build();
  }

  @Override
  public void close() throws IOException {
    detector.close();
  }
}
